analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantConditionalAroundAssignment.cs

```diff
@@ -36,58 +38,82 @@ public sealed class RedundantConditionalAroundAssignment : SonarDiagnosticAnalyz
 
-        private static readonly DiagnosticDescriptor rule =
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(rule);
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
 
         protected override void Initialize(SonarAnalysisContext context)
         {
-            context.RegisterSyntaxNodeActionInNonGenerated(
-                c =>
+            context.RegisterSyntaxNodeActionInNonGenerated(UselessConditionIfStatement, SyntaxKind.IfStatement);
+            context.RegisterSyntaxNodeActionInNonGenerated(UselessConditionSwitchExpression, SyntaxKindEx.SwitchExpression);
+        }
+
+        private static void UselessConditionIfStatement(SyntaxNodeAnalysisContext c)
+        {
+            var ifStatement = (IfStatementSyntax)c.Node;
+
+            if (ifStatement.Else != null
+                || ifStatement.Parent is ElseClauseSyntax
+                || (ifStatement.FirstAncestorOrSelf<AccessorDeclarationSyntax>()?.IsKind(SyntaxKind.SetAccessorDeclaration) ?? false)
+                || !TryGetNotEqualsCondition(ifStatement, out var condition)
+                || !TryGetSingleAssignment(ifStatement, out var assignment))
+            {
+                return;
+            }
+
+            var expression1Condition = condition.Left?.RemoveParentheses();
+            var expression2Condition = condition.Right?.RemoveParentheses();
+            var expression1Assignment = assignment.Left?.RemoveParentheses();
+            var expression2Assignment = assignment.Right?.RemoveParentheses();
+
+            if (!AreMatchingExpressions(expression1Condition, expression2Condition, expression2Assignment, expression1Assignment)
+                && !AreMatchingExpressions(expression1Condition, expression2Condition, expression1Assignment, expression2Assignment))
+            {
+                return;
+            }
+
+            if (!(c.SemanticModel.GetSymbolInfo(assignment.Left).Symbol is IPropertySymbol))
+            {
+                c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, condition.GetLocation()));
+            }
+        }
+
+        private static void UselessConditionSwitchExpression(SyntaxNodeAnalysisContext c)
+        {
+            var switchExpression = (SwitchExpressionSyntaxWrapper)c.Node;
+
+            if (!(switchExpression.SyntaxNode.GetFirstNonParenthesizedParent() is AssignmentExpressionSyntax parentExpression)
+                || !CSharpEquivalenceChecker.AreEquivalent(parentExpression.Left, switchExpression.GoverningExpression))
```

Comment:
Do we need this equivalence check? I think because of it we do not handle cases like:
```
y = x switch
{
    _ => 5       // Noncompliant
};
```